File: TESTING/MATGEN/zlatm2.f
      COMPLEX*16   FUNCTION ZLATM2( M, N, I, J, KL, KU, IDIST, ISEED, D, IGRADE, DL, DR, IPVTNG, IWORK, SPARSE )

*  -- LAPACK auxiliary routine --
*  -- LAPACK is a software package provided by Univ. of Tennessee,    --
*  -- Univ. of California Berkeley, Univ. of Colorado Denver and NAG Ltd..--

      // .. Scalar Arguments ..

      int                I, IDIST, IGRADE, IPVTNG, J, KL, KU, M, N;
      double             SPARSE;
      // ..

      // .. Array Arguments ..

      int                ISEED( 4 ), IWORK( * );
      COMPLEX*16         D( * ), DL( * ), DR( * )
      // ..

*  =====================================================================

      // .. Parameters ..

      COMPLEX*16         CZERO
      const              CZERO = ( 0.0D0, 0.0D0 ) ;
      double             ZERO;
      const              ZERO = 0.0D0 ;
      // ..

      // .. Local Scalars ..

      int                ISUB, JSUB;
      COMPLEX*16         CTEMP
      // ..

      // .. External Functions ..

      double             DLARAN;
      COMPLEX*16         ZLARND
      // EXTERNAL DLARAN, ZLARND
      // ..

      // .. Intrinsic Functions ..

      // INTRINSIC DCONJG
      // ..

*-----------------------------------------------------------------------

      // .. Executable Statements ..


      // Check for I and J in range

      IF( I.LT.1 .OR. I.GT.M .OR. J.LT.1 .OR. J.GT.N ) THEN
         ZLATM2 = CZERO
         RETURN
      END IF

      // Check for banding

      IF( J.GT.I+KU .OR. J.LT.I-KL ) THEN
         ZLATM2 = CZERO
         RETURN
      END IF

      // Check for sparsity

      IF( SPARSE.GT.ZERO ) THEN
         IF( DLARAN( ISEED ).LT.SPARSE ) THEN
            ZLATM2 = CZERO
            RETURN
         END IF
      END IF

      // Compute subscripts depending on IPVTNG

      IF( IPVTNG.EQ.0 ) THEN
         ISUB = I
         JSUB = J
      ELSE IF( IPVTNG.EQ.1 ) THEN
         ISUB = IWORK( I )
         JSUB = J
      ELSE IF( IPVTNG.EQ.2 ) THEN
         ISUB = I
         JSUB = IWORK( J )
      ELSE IF( IPVTNG.EQ.3 ) THEN
         ISUB = IWORK( I )
         JSUB = IWORK( J )
      END IF

      // Compute entry and grade it according to IGRADE

      IF( ISUB.EQ.JSUB ) THEN
         CTEMP = D( ISUB )
      ELSE
         CTEMP = ZLARND( IDIST, ISEED )
      END IF
      IF( IGRADE.EQ.1 ) THEN
         CTEMP = CTEMP*DL( ISUB )
      ELSE IF( IGRADE.EQ.2 ) THEN
         CTEMP = CTEMP*DR( JSUB )
      ELSE IF( IGRADE.EQ.3 ) THEN
         CTEMP = CTEMP*DL( ISUB )*DR( JSUB )
      ELSE IF( IGRADE.EQ.4 .AND. ISUB.NE.JSUB ) THEN
         CTEMP = CTEMP*DL( ISUB ) / DL( JSUB )
      ELSE IF( IGRADE.EQ.5 ) THEN
         CTEMP = CTEMP*DL( ISUB )*DCONJG( DL( JSUB ) )
      ELSE IF( IGRADE.EQ.6 ) THEN
         CTEMP = CTEMP*DL( ISUB )*DL( JSUB )
      END IF
      ZLATM2 = CTEMP
      RETURN

      // End of ZLATM2

      }

File: test/matgen/zlatm2.dart
// Copyright (c) 2024 Guilherme Lepsch. All rights reserved. Use of this
// source code is governed by a BSD-style license that can be found in the
// [LICENSE file](https://github.com/lepsch/dart_lapack/blob/main/LICENSE).

import 'package:lapack/src/complex.dart';
import 'package:lapack/src/matrix.dart';

import 'dlaran.dart';
import 'zlarnd.dart';

Complex zlatm2(
  final int M,
  final int N,
  final int I,
  final int J,
  final int KL,
  final int KU,
  final int IDIST,
  final Array<int> ISEED_,
  final Array<Complex> D_,
  final int IGRADE,
  final Array<Complex> DL_,
  final Array<Complex> DR_,
  final int IPVTNG,
  final Array<int> IWORK_,
  final double SPARSE,
) {
// -- LAPACK auxiliary routine --
// -- LAPACK is a software package provided by Univ. of Tennessee,    --
// -- Univ. of California Berkeley, Univ. of Colorado Denver and NAG Ltd..--
  final ISEED = ISEED_.having();
  final IWORK = IWORK_.having();
  final D = D_.having();
  final DL = DL_.having();
  final DR = DR_.having();
  const ZERO = 0.0;
  int ISUB = 0, JSUB = 0;
  Complex CTEMP;

  // Check for I and J in range

  if (I < 1 || I > M || J < 1 || J > N) {
    return Complex.zero;
  }

  // Check for banding

  if (J > I + KU || J < I - KL) {
    return Complex.zero;
  }

  // Check for sparsity

  if (SPARSE > ZERO) {
    if (dlaran(ISEED) < SPARSE) {
      return Complex.zero;
    }
  }

  // Compute subscripts depending on IPVTNG

  if (IPVTNG == 0) {
    ISUB = I;
    JSUB = J;
  } else if (IPVTNG == 1) {
    ISUB = IWORK[I];
    JSUB = J;
  } else if (IPVTNG == 2) {
    ISUB = I;
    JSUB = IWORK[J];
  } else if (IPVTNG == 3) {
    ISUB = IWORK[I];
    JSUB = IWORK[J];
  }

  // Compute entry and grade it according to IGRADE

  if (ISUB == JSUB) {
    CTEMP = D[ISUB];
  } else {
    CTEMP = zlarnd(IDIST, ISEED);
  }
  if (IGRADE == 1) {
    CTEMP *= DL[ISUB];
  } else if (IGRADE == 2) {
    CTEMP *= DR[JSUB];
  } else if (IGRADE == 3) {
    CTEMP *= DL[ISUB] * DR[JSUB];
  } else if (IGRADE == 4 && ISUB != JSUB) {
    CTEMP *= DL[ISUB] / DL[JSUB];
  } else if (IGRADE == 5) {
    CTEMP *= DL[ISUB] * DL[JSUB].conjugate();
  } else if (IGRADE == 6) {
    CTEMP *= DL[ISUB] * DL[JSUB];
  }
  return CTEMP;
}
